### PR TITLE
enhanced gateway api topology

### PR DESCRIPTION
**Pull request by eguzki**
### What

* new topology option `WithAcceptedRoutesLinkedOnly()`: only accepted routes will be linked to the gateway
* new topology option `WithProgrammedGatewaysOnly()`: only programmed (valid) gateways can have routes and attached policies. Not programmed gateways will be isolated nodes in the topology.
* new public method `Policies()` that returns all policies included in the topology. Returned objects are `PolicyNodes` that allow to jump to topology nodes targeted by the policy (aka `PolicyTargetNode`).

Work being done as part of https://github.com/Kuadrant/kuadrant-operator/issues/530

This is part of a series of PR's that will end with a new controller that will only reconcile limitador limits configuration.

### CODE REVIEW

File: pkg/rlptools/overrides.go [hunk 1]
```diff
@@ -37,7 +39,8 @@ func ApplyOverrides(topology *kuadrantgatewayapi.Topology, gateway *gatewayapiv1
 
 	for _, route := range topology.Routes() {
 		if !slices.Contains(kuadrantgatewayapi.GetRouteAcceptedGatewayParentKeys(route.HTTPRoute), client.ObjectKeyFromObject(gateway)) {
-			overriddenPolicies = append(overriddenPolicies, route.AttachedPolicies()...)
+			routePolicies := utils.Map(route.AttachedPolicies(), func(pNode kuadrantgatewayapi.PolicyNode) kuadrantgatewayapi.Policy { return pNode.Policy })
```
Comment (guicassolato) [inline pkg/rlptools/overrides.go:42]: TO avoid adding another dependency:

```suggestion
			routePolicies := lo.Map(route.AttachedPolicies(), func(pNode kuadrantgatewayapi.PolicyNode, _ int) kuadrantgatewayapi.Policy { return pNode.Policy })
```

File: pkg/library/gatewayapi/topology.go [hunk 2]
```diff
@@ -18,81 +18,192 @@ const (
...
+	programmedGatewaysOnly bool
+	linkAcceptedRoutesOnly bool
 }
 
 // TopologyOpts allows to manipulate topologyOptions.
 type TopologyOpts func(*topologyOptions)
 
+func WithAcceptedRoutesLinkedOnly() TopologyOpts {
+	return func(o *topologyOptions) {
+		o.linkAcceptedRoutesOnly = true
```
Comment (guicassolato) [inline pkg/library/gatewayapi/topology.go:197]: I guess we don't ever want non-accepted routes in the topology as we may occasionally want non-programmed gateways? I actually wonder what would be the use case for adding non-programmed gateways.

File: controllers/ratelimitpolicy_enforced_status_controller.go [hunk 3]
```diff
@@ -200,6 +200,8 @@ func (r *RateLimitPolicyEnforcedStatusReconciler) buildTopology(ctx context.Cont
 	}
 
 	return kuadrantgatewayapi.NewTopology(
+		kuadrantgatewayapi.WithAcceptedRoutesLinkedOnly(),
+		kuadrantgatewayapi.WithProgrammedGatewaysOnly(true),
```
Comment (guicassolato) [inline controllers/ratelimitpolicy_enforced_status_controller.go:204]: I understand a gateway that transitions from `Programmed: True` to `Programmed: False` will NOT get the statuses of its affecting policies updated then?

I'm thinking this line invoked for a gateway that has not been added to the topology: https://github.com/Kuadrant/kuadrant-operator/blob/4b9d60e49eae89460a963b9395f1e92a8f837332/controllers/ratelimitpolicy_enforced_status_controller.go#L60

File: pkg/library/gatewayapi/utils.go [hunk 4]
```diff
@@ -165,3 +165,23 @@ func IsGatewayAPIInstalled(restMapper meta.RESTMapper) (bool, error) {
 
 	return false, err
 }
+
+// GetGatewayParentRefsFromRoute returns the list of parentRefs that are Gateway typed
+func GetGatewayParentRefsFromRoute(route *gatewayapiv1.HTTPRoute) []gatewayapiv1.ParentReference {
+	if route == nil {
+		return nil
+	}
+	return utils.Filter(route.Spec.ParentRefs, IsParentGateway)
```
[the remark this answers is not in the document]
Comment (eguzki) [reply pkg/library/gatewayapi/utils.go:174]: I do not see any reason to change when `utils.Filter()` does what it is expected to do. Anything I could miss?

File: pkg/rlptools/overrides.go (hunk 1 above)
Comment (eguzki) [reply pkg/rlptools/overrides.go:42]: same as https://github.com/Kuadrant/kuadrant-operator/pull/751#discussion_r1672144674

File: pkg/library/gatewayapi/topology.go (hunk 2 above)
Comment (eguzki) [reply pkg/library/gatewayapi/topology.go:197]: there are some use cases I found that non-accepted routes should be linked to parent gateways: mappers. It allows getting events for "disconnected" policies, which status can be affected from the "disconnection".

File: controllers/ratelimitpolicy_enforced_status_controller.go (hunk 3 above)
Comment (eguzki) [reply controllers/ratelimitpolicy_enforced_status_controller.go:204]: good catch, thanks!

File: pkg/library/gatewayapi/topology.go (hunk 2 above)
Comment (eguzki) [reply pkg/library/gatewayapi/topology.go:197]: >  I actually wonder what would be the use case for adding non-programmed gateways.

this is a use case https://github.com/Kuadrant/kuadrant-operator/pull/751#discussion_r1671913835

File: controllers/ratelimitpolicy_enforced_status_controller.go (hunk 3 above)
Comment (eguzki) [reply controllers/ratelimitpolicy_enforced_status_controller.go]: fixed

File: pkg/rlptools/overrides.go (hunk 1 above)
Comment (eguzki) [reply pkg/rlptools/overrides.go:42]: Maybe, for other PR, we could remove all `utils.Map` and `utils.Filter` and rewrite using [github.com/samber/lo](https://pkg.go.dev/github.com/samber/lo) (or something else). Then we can remove `utils` code and have the external dependency only.

File: pkg/library/gatewayapi/utils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (guicassolato) [reply pkg/library/gatewayapi/utils.go:200]: `github.com/samber/lo` is immensely popular and IMO very good at this kind of generic widely used functions. Wherever we can leverage something like this from the community – provided it suits our needs of course – instead of keeping our own home-made version of the same thing, I personally would be in favour of.

At the same time, I don't want to create a fuss. I understand this is not the main goal of this PR. Just something to think about.

File: pkg/library/gatewayapi/topology.go (hunk 2 above)
Comment (guicassolato) [reply pkg/library/gatewayapi/topology.go:197]: > there are some use cases I found that non-accepted routes should be linked to parent gateways

So why not giving the option here then?

Comment (guicassolato) [reply pkg/library/gatewayapi/topology.go:197]: Sorry. To clarify...

My point is that we have `WithProgrammedGatewaysOnly(bool)` for when we want to enforce programmed gateways only, i.e. `WithProgrammedGatewaysOnly(true)`. I don't see why one would use `WithProgrammedGatewaysOnly(false)` though – versus simply not calling `WithProgrammedGatewaysOnly` at all.

I guess that was your thinking for `WithAcceptedRoutesLinkedOnly`. Nobody will ever need `WithAcceptedRoutesLinkedOnly(false)`. The effect in this case is virtually the same as not using `WithAcceptedRoutesLinkedOnly`, right?

So why not designing `WithProgrammedGatewaysOnly` the same way? If there's a reason I struggling to see, then there may be for `WithAcceptedRoutesLinkedOnly` too perhaps?

IOW, why are the signatures of these 2 functions different?

Comment (eguzki) [reply pkg/library/gatewayapi/topology.go:197]: Makes all the sense. Fixing it

Comment (eguzki) [reply pkg/library/gatewayapi/topology.go:197]: done

File: controllers/authpolicy_status.go [hunk 5]
```diff
@@ -169,6 +169,8 @@ func (r *AuthPolicyReconciler) generateTopology(ctx context.Context) (*kuadrantg
 	})
 
 	return kuadrantgatewayapi.NewTopology(
+		kuadrantgatewayapi.WithAcceptedRoutesLinkedOnly(),
+		kuadrantgatewayapi.WithProgrammedGatewaysOnly(true),
```
[the remark this answers is not in the document]
Comment (guicassolato) [reply controllers/authpolicy_status.go]: BTW, the described problem is not introduced by this PR, but neither it's fixed by it though. In fact, I'm afraid the change may even aggravate the problem.

The more fundamental problem is of course that, in this case, the topology is only marginal to the decisions we make. Most of the business logic for reconciling AuthPolicies still won't leverage the topology graph built by this function, while some already do.

By trying to fix _only_ the topology graph, we're effectively providing two views of world, thus splitting the business logic into the portion that creates configuration based on one reality and another based on a different reality.

I fear that, by trying to do the right thing here, we're actually bringing the system to an inconsistent state.

[the remark this answers is not in the document]
Comment (eguzki) [reply controllers/authpolicy_status.go]: :exploding_head: 

I am confused. I will kindly ask you to elaborate more on this because I do not get it. 

I will try to clarify some of your statements:

> BTW, the described problem is not introduced by this PR, but neither it's fixed by it though.

This PR does not pretend to change a bit the behavior of any of the controllers. All the tests pass. There is no problem being described and thus, no problem being fixed.

This PR is adding few more options to the DAG topology, yet keeping the controllers behavior as they are. The intention is to add more tools/capabilities on the topology lib to enable following up PR's introducing the Limitador  limits controller. 

The authcontroller in current `main` branch is building a topology with only programmed gateways and linking only accepted routes. So is doing the code with the new changes in this PR. Nothing changes. And if something changes, let me know and I will fix it. 

 > In fact, I'm afraid the change may even aggravate the problem.

First of all, what problem are you talking about? 

Configuration for "programmed gateways only" option was introduced by https://github.com/Kuadrant/kuadrant-operator/pull/715. This PR is adding (among other public methods) yet another configuration option which is opt out from "link accepted routes only". As you know, currently the topology has [hard-coded](https://github.com/Kuadrant/kuadrant-operator/blob/c8d02d0437956dc66c14f4a04327c435a1b66fa3/pkg/library/gatewayapi/topology.go#L198) linking only accepted routes. This may not be always desirable. This feature is not needed now, but will be needed by the limits controller. The reasoning being that the controller will  need a topology in its mappers which routes are linked to the gateway no matter they are accepted or not. But this is something I can explain further when I open PR for the limits controller.  

I do not see any problem being aggravated. Again, this is not changing any controller behavior. And if it does, I will fix it.

> By trying to fix only the topology graph, we're effectively providing two views of world, thus splitting the business logic into the portion that creates configuration based on one reality and another based on a different reality.

This is by no means fixing the topology. AFAIK, there is no issue with the current topology for the needs of the current controllers. This is about adding more configurability to the topology. Some business logic requires a type of topology with not programmed gateways as well as linked routes that have not been accepted. Specifically I am thinking on the mappers. Mappers need to walk the topology. And having disconnections because routes are not accepted prunes events to reconcile kuadrant policies.

[the remark this answers is not in the document]
Comment (guicassolato) [reply controllers/authpolicy_status.go]: > Configuration for "programmed gateways only" option was introduced by #715. […] As you know, currently the topology has [hard-coded](https://github.com/Kuadrant/kuadrant-operator/blob/c8d02d0437956dc66c14f4a04327c435a1b66fa3/pkg/library/gatewayapi/topology.go#L198) linking only accepted routes.

This indeed confirms the PR does not aggravate the problem, and it's enough for me to approve it. Thanks for explaining it and for collecting the evidences in the code to support it, @eguzki .

To elaborate on my main criticism, which I still stand for...

<table>
<tr>
<td>
<b>Disclaimer for all readers:</b> please try to read in between the lines and focus on the generalized form of the problem I'm trying to describe, rather than the individual examples I use.
</td>
</tr>
</table>

I missed #715. I think this is a very good example of how a well-intended change, that we needed indeed and that, looked individually, made our code better than before, but, at the same time, if observed holistically, emerges nonetheless as a change that creates a problem (or aggravates one) we didn't have before (or that simply had a smaller surface before.)

We spotted we were treating non-programmed gateways as if they were ready and thus we were signaling the policies configured on those gateways as enforced. This was clearly a bug and had to be fixed. We have tons of code that implement business logic based on a "view of the world" provided by this topology struct. We made that topology struct to only account for programmed gateways. Done. Problem solved. We fixed all those business logic with one move.

However, unfortunately there are controllers (possibly all of them AFAIK!) that mix business logic based on the topology struct with business logic that does not use the topology struct, where sometimes the status of the Gateways and HTTPRoutes are properly taken into account, sometimes not. (I mentioned the reconciliation of AuthConfigs before, but look at [`ComputeGatewayDiff`](https://github.com/Kuadrant/kuadrant-operator/blob/c8d02d0437956dc66c14f4a04327c435a1b66fa3/pkg/library/reconcilers/gateway_diffs.go#L27) for plenty of more examples.)

With the topology struct now doing what seems to be the right thing, we created two "views of the world" – one that we believe is right and another one that may or may not be right. These two views (as well as the decisions based on them) are often intertwined together, leading to a situation where we can no longer say _"Kuadrant overlooks the state of the gateways. As a user, it's up to you to make sure to bring all your gateways to a programmed state. Until then, do not trust status of Kuadrant policies."_ Unfortunately, neither we can say today, after #715, _"No worries! Kuadrant makes sure to always configure and report status of your policies right, according to the state of the gateways."_ Between the two statements, obviously we want to issue the latter, but if that's not true, then we're probably better off with the former IMO.

I used "programmed gateways" as example, but we can run over again replacing it with "accepted routes", or ["listener status"](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.ListenerStatus), and possibly others that do not occur to me right now. The important part is, while we are not consistent about the views of the world that the code bases its decisions upon, it becomes very hard to debug behaviours such as:
- why a policy with status enforced doesn't seem to be enforced at all
- why an override that seems not have been accepted appears to be taken over my route-level rules (or the other way around)
- etc

To sum up, let's be careful with changes that seem to be improving the code but until they are rolled out completely they may actually put us (as well as our users) in a difficult position. This is not the case of this PR as I'm now convinced, and neither it is a negative criticism to the DAG/topology approach in general. Rather, I'm concerned with how we roll out these changes.

**Review by guicassolato — APPROVED** [review]
